cmd/cleanup-export/main.go

```diff
@@ -17,35 +17,64 @@ package main
 
 import (
 	"context"
-	"log"
+	"fmt"
 	"net/http"
-
-	"go.opencensus.io/plugin/ochttp"
+	"time"
 
 	"github.com/google/exposure-notifications-server/internal/cleanup"
+	"github.com/google/exposure-notifications-server/internal/interrupt"
 	"github.com/google/exposure-notifications-server/internal/logging"
 	_ "github.com/google/exposure-notifications-server/internal/observability"
+	"github.com/google/exposure-notifications-server/internal/server"
 	"github.com/google/exposure-notifications-server/internal/setup"
 )
 
 func main() {
-	ctx := context.Background()
+	ctx, done := interrupt.Context()
+	defer done()
+
+	if err := realMain(ctx); err != nil {
+		logger := logging.FromContext(ctx)
+		logger.Fatal(err)
+	}
+}
+
+func realMain(ctx context.Context) error {
 	logger := logging.FromContext(ctx)
 
 	var config cleanup.Config
 	env, closer, err := setup.Setup(ctx, &config)
 	if err != nil {
-		logger.Fatalf("setup.Setup: %v", err)
+		return fmt.Errorf("setup.Setup: %w", err)
 	}
 	defer closer()
 
 	handler, err := cleanup.NewExportHandler(&config, env)
 	if err != nil {
-		logger.Fatalf("cleanup.NewExportHandler: %v", err)
+		return fmt.Errorf("cleanup.NewExportHandler: %w", err)
 	}
+
 	mux := http.NewServeMux()
 	mux.Handle("/", handler)
-	logger.Infof("starting export cleanup server on :%s", config.Port)
-	instrumentedHandler := &ochttp.Handler{Handler: mux}
-	log.Fatal(http.ListenAndServe(":"+config.Port, instrumentedHandler))
+
+	server := server.New(config.Port, mux)
```

Comment:
I'm open to unifying `New` and `Start` into a single API, since there's no _real_ value in `New` since all the fields are private. 

I also thought about an API where the context provided to `server.Start` is used to stop it:

```go
ctx, done := context.WithCancel(ctx)
if err := server.Start(ctx); err != nil {
  return err
}

<-ctx
done()
```

but it felt too magical, and it's much harder to return any errors that occur while stopping back to the main thread.

Reply:
I also thought about a `server.Wait(ctx)`, which made a bit more sense. I'm open to other designs here that reduce duplication.

Reply:
seems like everything here on down is "common main" should we pass the mux to a "run server" method

Reply:
usually it's `server.Serve(ctx)` that blocks until it shuts down gracefully. that way you don't need a stop method.

I think separate constructor is fine
lets you separate serve errors from init errors

Reply:
@adg thanks for the feedback. The problem we're trying to solve is intercepting a signal (like SIGTERM/SIGINT) and having that close the server. `interrupt.Context()` creates a context that closes on various signals, but I'm seeking advice on the best way to "plumb" that through to the servers.

Reply:
Right, I can see that. I'm saying have a single Serve method that takes a ctx and have it block until that context is done, then it does the graceful shutdown (using a background context, with timeout, for any shutdown-related things that require a context), then returns. Then the last line of your main function can be `return server.Serve(ctx)`

Reply:
@adg @mikehelmick updated. PTAL.